cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        .setNamespaceId(streamId.getNamespaceId())
+        .setCategory(Constants.Notification.Stream.STREAM_FEED_CATEGORY)
+        .setName(String.format("%sSize", streamId.getName()))
+        .build();
+    }
+
+    /**
+     * Cancel the currently scheduled stream size polling task, and reschedule one for later.
+     */
+    private void cancelPollingAndScheduleNext() {
+      // This method might be called from the call to #received defined in the below Runnable - in which case
+      // this scheduledPolling would in fact be running. Hence we don't want to interrupt the running task
+      if (scheduledPolling != null) {
+        scheduledPolling.cancel(false);
+      }
+
+      // Regardless of whether cancelling was successful, we still want to schedule the next polling
+      scheduledPolling = streamPollingExecutor.schedule(createPollingRunnable(), pollingDelay, TimeUnit.SECONDS);
+    }
+
+    /**
+     * @return a runnable that uses the {@link StreamAdmin} to poll the stream size, and creates a fake notification
+     *         with that size, so that this information can be treated as if it came from a real notification.
+     */
+    private Runnable createPollingRunnable() {
+      return new Runnable() {
+        @Override
+        public void run() {
+          // We only perform polling if at least one scheduleTask is active
+          if (activeTasks.get() > 0) {
+            long size = pollStream();
+
+            // We don't need a notification context here
+            received(new StreamSizeNotification(System.currentTimeMillis(), size), null);
+          }
+        }
+      };
+    }
+
+    /**
+     * @return size of the stream queried directly from the file system
+     */
+    private long pollStream() {
+      try {
+        // Note we can't store the stream config, because its generation might change at every moment
+        return streamAdmin.fetchStreamSize(streamAdmin.getConfig(streamId));
+      } catch (IOException e) {
+        LOG.error("Could not poll size for stream {}", streamId);
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private final StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private boolean running;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
```

Comment:
For now it is, until the persistent store is implemented (coming in a future PR).